### DISCUSSION TITLE
How to set a checkbox to true?

### DISCUSSION THREAD

**Discussion by ColeBlender** [py-pdf/pypdf#2004] [Q&A]
Once I get the appropriate key I can do this just fine with text boxes:
`writer.update_page_form_field_values(writer.pages[i], {key: "Jeff Willams"})`

How do I do it with checkboxes? I've tried all of following with no success:
`writer.update_page_form_field_values(writer.pages[i], {key: True})
writer.update_page_form_field_values(writer.pages[i], {key: "True"})
writer.update_page_form_field_values(writer.pages[i], {key: "true"})
writer.update_page_form_field_values(writer.pages[i], {key: "On"})
writer.update_page_form_field_values(writer.pages[i], {key: "/On"})
writer.update_page_form_field_values(writer.pages[i], {key: "Yes"})
writer.update_page_form_field_values(writer.pages[i], {key: "/Yes"})`

Is there a different method for this? Am I using the wrong value? Is this even possible using this library?

**Comment by pubpub-zz** [comment]
a) can you confirm you are on the latest version from pypdf
b) the active state differs between checkboxes. To identify the states, you should get the informations using `get_fields()` and look for `["/_States_"]` entry from the field to identify it.

**Reply by ColeBlender** [reply]
pypdf==3.12.1
Yes that's exactly what I've done. I'm able to see what the current value is. My question is how to change it. Is there a method designed for that or do I have to go and manually change it? If that's the case then do I just update where I see ["/_States_"] or is that just a reference to some other place that I have to go update?

**Comment by pubpub-zz** [comment]
from the test suite:
```
import pypdf
rrr = pypdf.PdfReader("e:/PyPDF4.git/PyPDF2/resources/FormTestFromOo.pdf")
rrr.get_fields()["CheckBox1"] 
# returns: {'/T': 'CheckBox1', '/FT': '/Btn', '/TU': 'help for CB', '/V': '/Off', '/DV': '/Off', '/_States_': ['/Yes', '/Off']}

www = pypdf.PdfWriter(clone_from=rrr)
www.update_page_form_field_values(
        www.pages[0],
        {
            "CheckBox1": "/Yes",
            "Text1": "mon Text1",
            "Text2": "ligne1\nligne2",
            "RadioGroup1": "/2",
            "RdoS1": "/",
            "Combo1": "!!monCombo!!",
            "Liste1": "Liste2",
            "Liste2": ["Lst1", "Lst3"],
            "DropList1": "DropListe3",
        },
        auto_regenerate=False,
    )
```
and the result:
![image](https://github.com/py-pdf/pypdf/assets/4083478/fc7235af-c801-45ca-8e80-50d800cc54d0)

**Reply by ColeBlender** [reply]
I'm not able to get it to work. Maybe the pdf is different somehow. Here's the link if you want to give it a try: [https://www.uscis.gov/sites/default/files/document/forms/i-589.pdf](url)
Here's my code also:
`pdf = "app/pdf/i-589.pdf"
reader = PdfReader(pdf)
writer = PdfWriter()
writer.append(reader)

total_pages = len(reader.pages)
fields = reader.get_fields()

for key in fields:
    if is_safe_to_fill(key):
        for n in range(total_pages):
            if is_check_box(fields[key]):
                states = fields[key]["/_States_"]
                # state = [s for s in states if s != "/Off"][0]
                # new_states = [state, "/Off"]
                # print(fields[key])
                writer.update_page_form_field_values(writer.pages[n], {key: "/Y"}, auto_regenerate=False)
            else:
                writer.update_page_form_field_values(writer.pages[n], {key: "TEST"})

with open("app/pdf/filled-out.pdf", "wb") as output_stream:
    writer.write(output_stream)`

**Reply by ColeBlender** [reply]
def is_safe_to_fill(key):
    if "#subform" in key:
        text = key.split("subform[")[1]
        if text.split("]")[1] == "":
            return False
        return True
    else:
        return False


def is_check_box(value):
    return "/_States_" in value

**Reply by pubpub-zz** [reply]
I may not have some problem to understand completely your code. but this is my proposal:
 ```
pdf = "app/pdf/i-589.pdf"
reader = PdfReader(pdf)
writer = PdfWriter(clone_from=reader)

fields = reader.get_fields()
for key,field in fields.items():
    if is_safe_to_fill(key):
        p = field.indirect_reference.get_object().get("/P")
        if p is None:
            continue
        n = r.get_page_number(p)
        if is_check_box(field):
            states = field["/_States_"]
            state = [s for s in states if s != "/Off"][0]
            writer.update_page_form_field_values(writer.pages[n], {key: state}, auto_regenerate=False)
        else:
            writer.update_page_form_field_values(writer.pages[n], {key: "TEST"})

writer.write("app/pdf/filled-out.pdf")
```

**Reply by KyleSCraig** [reply]
@pubpub-zz Thanks so much for this code here. I'm trying to fill PDFs that have a unique **On** state for checkboxes and this is so much nicer than working out what they are. Also I believe I saw you added /\_States\_ without which I wouldn't have even been able to see what the **On** state needed to be.
